### PR TITLE
Git actions: direct commit mode action execution via kubeturbo

### PR DESCRIPTION
**Pull request by irfanurrehman**
**Intent**
This PR implements support for actions executed on applications managed by gitops pipeline specific to `argoCD` as the CD tool and `github` as source of truth.
This implements https://vmturbo.atlassian.net/browse/OM-81584 as the story.

**Background**
There is a considerable number of users which rely on the gitops based controllers and a source of truth, eg a git repo to manage the statue of resources in their clusters. Turbo as of now cannot execute resource resize and horizontal scale actions on such environments. Turbo can (and does) apply the changes spec on the workload controller resource that exists in the cluster, but as its managed by a gitops based controller, its immediately synced with the source of truth and reverted back to the original. This PR implements basic support using which the actions result will be updated by pushing a PR to the source of truth. This PR assumes that as of now only supported types will be a git repo with yaml files stored in it.

**Implementation**
This work is done on top of earlier work in #615. The earlier work tried using an external `Change Request` and an associated controller to update the source of truth, Some of the code from that work is retained (`pkg/action/executor/change_request_controller.go`) but not used anywhere. If we decide that we don't want to use the CR approach ever, we can remove the file.

The implementation includes:
- Discovering `argoCD` applications as business applications in turbo supply chain
- Storing the details about the managing argoCD app (or turbo business app) as a set of properties on the corresponding workload controller. This detail is then made available on the workload controller in the action DTO.
- Implements an action executor which can update the source of truth via a `githubManager`, with an intent that other managers (eg `bitbucketManager`) can be implemented on similar lines.
- The action executor identifies the file matching the resource name found in k8s cluster to the name specified in the yaml stored in github, patches the same with the update requested in the turbo action and updates that as a new commit on the specified branch.
- This also introduces 4 new kubeturbo parameters:
```
      --git-email string                           The email to be used to push changes to git.
      --git-secret-name string                     The name of the secret which holds the git credentials.
      --git-secret-namespace string                The namespace of the secret which holds the git credentials.
      --git-username string                        The user name to be used to push changes to git.
```

This is a fairly large piece of work and review can be looked at based on commits. We can also have an online review if thats more favourable.

**tests**
- ArgoCD app discovered as a business app
```
irf@Irfans-MacBook-Pro-2 kubeturbo % k get applications -A
NAMESPACE   NAME               SYNC STATUS   HEALTH STATUS
argocd      gitops-multifile   Synced        Healthy
```
image

- Action execution on a workload managed by argoCD succeeds
```
I0321 10:22:36.280107       1 action_handler.go:362] Received an action RIGHT_SIZE for entity WORKLOAD_CONTROLLER [beekman-change-reconciler]
I0321 10:22:36.280163       1 action_handler.go:179] Now wait for action result
I0321 10:22:36.280406       1 remote_mediation_client.go:488] Sent action progress for 109.
I0321 10:22:36.280453       1 remote_mediation_client.go:239] [probeCallback] received response on probe channel 0xc000e81380

I0321 10:22:36.280536       1 client_protobuf_endpoint.go:78] [ServerRequestEndpoint] : Sending protobuf message
I0321 10:22:36.280853       1 client_websocket_transport.go:266] Successfully sent message on client transport
I0321 10:22:36.280880       1 remote_mediation_client.go:233] [probeCallback] waiting for probe responses
I0321 10:22:36.290820       1 resize_container.go:133] Resize beekman-change-reconciler VCPU Capacity to 100
I0321 10:22:36.290857       1 workload_controller_resizer.go:170] Begin to resize workload controller default/beekman-change-reconciler.
I0321 10:22:36.290889       1 k8s_controller_updater.go:145] Begin to update Deployment default/beekman-change-reconciler
I0321 10:22:36.294906       1 k8s_controller_updater.go:195] Try to update resources for container indexes: 0, in pod template spec of Deployment default/beekman-change-reconciler .
I0321 10:22:36.294951       1 resize_container_util.go:50] Begin to update Capacity.
I0321 10:22:36.294967       1 resize_container_util.go:69] Try to update container beekman-1 resource limit from map[cpu:{i:{value:11 scale:-3} d:{Dec:<nil>} s:11m Format:DecimalSI} memory:{i:{value:29360128 scale:0} d:{Dec:<nil>} s:28Mi Format:BinarySI}] to map[cpu:{{100 -3} {<nil>} 100m DecimalSI} memory:{{29360128 0} {<nil>} 28Mi BinarySI}]
I0321 10:22:36.295008       1 resize_container_util.go:130] Container beekman-1 resources changed.
I0321 10:22:39.281174       1 remote_mediation_client.go:239] [probeCallback] received response on probe channel 0xc000e81380

I0321 10:22:39.281308       1 client_protobuf_endpoint.go:78] [ServerRequestEndpoint] : Sending protobuf message
I0321 10:22:39.281466       1 client_websocket_transport.go:266] Successfully sent message on client transport
I0321 10:22:39.281578       1 remote_mediation_client.go:233] [probeCallback] waiting for probe responses
I0321 10:22:39.281149       1 remote_mediation_client.go:488] Sent action progress for 109.
I0321 10:22:41.794086       1 k8s_controller_updater.go:163] Successfully updated Deployment default/beekman-change-reconciler
I0321 10:22:41.794148       1 go_util.go:59] [retry-1/3] success
I0321 10:22:41.794193       1 remote_mediation_client.go:469] Sent action response for 109.
```

image

image

```
irf@Irfans-MacBook-Pro-2 kubeturbo % k get pods
NAME                                         READY   STATUS        RESTARTS   AGE
beekman-change-reconciler-75ffc44bcd-vq6j8   1/1     Terminating   0          7h40m
beekman-change-reconciler-75ffc44bcd-wzpdb   1/1     Terminating   0          7h40m
beekman-change-reconciler-7b9bbfdb76-57r26   1/1     Running       0          18s
beekman-change-reconciler-7b9bbfdb76-8f4pv   1/1     Running       0          14s
beekman-one-more-75ffc44bcd-64kv7            1/1     Running       0          4h
```

Also updated a http based unit tester for the github flow code. Introduced few tests which can be extended.

### CODE REVIEW

File: pkg/action/executor/change_request_controller.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+		// TODO: Implement a wait loop here to retrieve the CR again and wait
+		// until the status is either "Complete" or "Failed".
+		// Once the retrieved status is complete, resetting the status to ""
+		// and updating the spec would mean the CR controller will try to push
+		// the changes again. We will also need to ensure that pushing the change
+		// does not mean creating multiple PRs if the previous one exists.
+
+		// TODO: Reset the status too, because if the status still remains
+		// 'completed' the change-reconciler would not do anything.
+		// Brainstorm a little more on this.
```
Comment (chlam4) [inline pkg/action/executor/change_request_controller.go:150]: It's tricky to synchronize on a mutable object.  I wonder if it's cleaner to create a new change request each time with a timestamp.  The change reconciler will populate the status as complete, failed or skipped (if a newer one exists before the older one is processed).  The change request controller can choose to delete a change request once it's fulfilled.

**Review by chlam4 — COMMENTED** [review]
Really nice work @irfanurrehman!  A few comments so far - lots of them were thoughts stimulated by your work.  I'll continue reviewing the rest tomorrow.

File: pkg/action/executor/k8s_controller_updater.go [hunk 2]
```diff
@@ -47,20 +49,34 @@ func newK8sControllerUpdaterViaPod(clusterScraper *cluster.ClusterScraper, pod *
 	if discoveryutil.IsOwnerInfoEmpty(ownerInfo) {
 		return nil, fmt.Errorf("pod %s/%s does not have controller", pod.Namespace, pod.Name)
 	}
-	return newK8sControllerUpdater(clusterScraper, ormClient, ownerInfo.Kind, ownerInfo.Name, pod.Name, pod.Namespace)
+	// TODO: For an action on a parent workload controller of this pod managed by gitops (argoCD) we will need
+	// Some information to be available on the entity we get in the action (pod here).
```
Comment (chlam4) [inline pkg/action/executor/k8s_controller_updater.go:53]: Nit: The starting "S" should be lower case.

**Review by chlam4 — APPROVED** [review]
Overall looks good to me as a PoV feature.  Thanks for the great work, Irfan!

File: cmd/kubeturbo/app/kubeturbo_builder.go [hunk 3]
```diff
@@ -204,6 +207,11 @@ func (s *VMTServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.containerUtilizationDataAggStrategy, "cnt-utilization-data-agg-strategy", agg.DefaultContainerUtilizationDataAggStrategy, "Container utilization data aggregation strategy.")
 	fs.StringVar(&s.containerUsageDataAggStrategy, "cnt-usage-data-agg-strategy", agg.DefaultContainerUsageDataAggStrategy, "Container usage data aggregation strategy.")
 	fs.IntVar(&s.readinessRetryThreshold, "readiness-retry-threshold", DefaultReadinessRetryThreshold, "When the pod readiness check fails, Kubeturbo will try readinessRetryThreshold times before giving up. Defaults to 60.")
+	// Flags for gitops based action execution
+	fs.StringVar(&s.gitConfig.GitSecretNamespace, "git-secret-namespace", "", "The namespace of the secret which holds the git credentials.")
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply cmd/kubeturbo/app/kubeturbo_builder.go:211]: Yes, that was the idea but I did miss the implementation for the same. Thanks for pointing this out. Updated!

File: pkg/action/executor/change_request_controller.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	CRVersion      = "v1alpha1"
+	CRResourceName = "changerequests"
+	CRKind         = "ChangeRequest"
+
+	// We currently use the CR related values as an annotation on the
+	// given workload controller. At some point we will need to move this
+	// to a more structured place, ie. a custom resource.
+	GitopsSourceAnnotationKey          = "turbonomic.io/gitops-source"
+	GitopsFilePathAnnotationKey        = "turbonomic.io/gitops-file-path"
+	GitopsBranchAnnotationKey          = "turbonomic.io/gitops-branch"
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/change_request_controller.go:28]: Not that I know of. Different tools (for example argo) can have their own annotations.

File: pkg/action/executor/change_request_controller.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+const (
+	CRGroup        = "turbonomic.io"
+	CRVersion      = "v1alpha1"
+	CRResourceName = "changerequests"
+	CRKind         = "ChangeRequest"
+
+	// We currently use the CR related values as an annotation on the
+	// given workload controller. At some point we will need to move this
+	// to a more structured place, ie. a custom resource.
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/change_request_controller.go:25]: This makes a lot of sense. I will create a story to capture this suggestion.

File: pkg/action/executor/change_request_controller.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+	// json representation of the obj
+	// The creation of ChangeRequest will be observed by the
+	// change reconciler which will update the source path
+	// with the new object
+	Update(replicas int64, podSpec map[string]interface{}) error
+}
+
+type cRControllerImpl struct {
+	dynClient  dynamic.Interface
+	obj        *unstructured.Unstructured
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/change_request_controller.go:45]: Removed the file as in [this comment](https://github.com/turbonomic/kubeturbo/pull/655#discussion_r847272590)

File: pkg/action/executor/change_request_controller.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+	}
+}
+
+func (cr *cRControllerImpl) Update(replicas int64, podSpec map[string]interface{}) error {
+	path, repo, revision, err := cr.getFieldsFromManagerApp()
+	if err != nil {
+		return err
+	}
+
+	cR := newCRResource(cr.obj.GetName(), cr.obj.GetNamespace())
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/change_request_controller.go:63]: Removed the file as in [this comment](https://github.com/turbonomic/kubeturbo/pull/655#discussion_r847272590)

File: pkg/action/executor/change_request_controller.go (hunk 1 above)
Comment (irfanurrehman) [reply pkg/action/executor/change_request_controller.go:150]: Removed the file as in [this comment](https://github.com/turbonomic/kubeturbo/pull/655#discussion_r847272590)

File: pkg/action/executor/github_manager.go [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+		user:        pathParts[1],
+		repo:        strings.TrimSuffix(pathParts[2], ".git"),
+		baseBranch:  baseBranch,
+		path:        path,
+		commitUser:  r.gitConfig.GitUsername,
+		commitEmail: r.gitConfig.GitUsername,
+	}
+
+	var patches []PatchItem
+	patches = append(patches,
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/github_manager.go:114]: Updated!

File: pkg/action/executor/github_manager.go [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+	return handler.updateRemote(r.obj.GetName(), patches)
+
+}
+
+func (r *GitManager) getAuthTokenFromSecret() (string, error) {
+	name := r.gitConfig.GitSecretName
+	namespace := r.gitConfig.GitSecretNamespace
+	if r.gitConfig.GitSecretName == "" {
+		return "", fmt.Errorf("secret name empty. It is necessary to get github auth token."+
+			"Updating workload controller %s", r)
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/github_manager.go:137]: Updated to below, please let me know if that also does not make sense.
```
		return "", fmt.Errorf("secret name found empty while updating workload controller %s. "+
			"It is necessary to get github auth token", r)
```

File: pkg/action/executor/github_manager.go [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+		return "", "", "", fmt.Errorf("required field path not found in manager app %s/%s: %v", app.GetNamespace(), app.GetName(), err)
+	}
+	repo, found, err = unstructured.NestedString(app.Object, "spec", "source", "repoURL")
+	if err != nil || !found {
+		return "", "", "", fmt.Errorf("required field repoURL not found in manager app %s/%s: %v", app.GetNamespace(), app.GetName(), err)
+	}
+	revision, found, err = unstructured.NestedString(app.Object, "spec", "source", "targetRevision")
+	if err != nil || !found {
+		return "", "", "", fmt.Errorf("required field targetRevision not found in manager app %s/%s: %v", app.GetNamespace(), app.GetName(), err)
+	}
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/github_manager.go:179]: Removed `change_request_controller.go` as in [this comment](https://github.com/turbonomic/kubeturbo/pull/655#discussion_r847272590)

File: pkg/action/executor/k8s_controller_updater.go (hunk 2 above)
Comment (irfanurrehman) [reply pkg/action/executor/k8s_controller_updater.go:53]: Updated!

File: pkg/action/executor/github_manager.go [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+
+	if obj.GetName() == name {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (g *GitHandler) fileContentFromDirContent(resName string, dirContent []*github.RepositoryContent) (string, string, error) {
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/github_manager.go:255]: I will ensure to update the details as part of my PR mode support work. Some amount of this code will get reused there and possibly refactored to support the PR mode. Please let me know if that is not ok.

File: pkg/action/executor/change_request_controller.go [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+	// given workload controller. At some point we will need to move this
+	// to a more structured place, ie. a custom resource.
+	GitopsSourceAnnotationKey          = "turbonomic.io/gitops-source"
+	GitopsFilePathAnnotationKey        = "turbonomic.io/gitops-file-path"
+	GitopsBranchAnnotationKey          = "turbonomic.io/gitops-branch"
+	GitopsSecretNameAnnotationKey      = "turbonomic.io/gitops-secret-name"
+	GitopsSecretNamespaceAnnotationKey = "turbonomic.io/gitops-secret-namespace"
+)
+
+type cRController interface {
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/change_request_controller.go:33]: Removed this file as in [this comment](https://github.com/turbonomic/kubeturbo/pull/655#discussion_r847272590)

File: pkg/action/executor/change_request_controller.go [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+	// Update will create a ChangeRequest resource with
+	// path from the obj annotation and payload as the
+	// json representation of the obj
+	// The creation of ChangeRequest will be observed by the
+	// change reconciler which will update the source path
+	// with the new object
+	Update(replicas int64, podSpec map[string]interface{}) error
+}
+
+type cRControllerImpl struct {
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/change_request_controller.go:43]: Removed this file as in [this comment](https://github.com/turbonomic/kubeturbo/pull/655#discussion_r847272590)

File: pkg/action/executor/change_request_controller.go [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	return outerErr
+}
+
+func (c *cRControllerImpl) getFieldsFromManagerApp() (string, string, string, error) {
+	if c.managerApp == nil {
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/change_request_controller.go:161]: Removed this file as in [this comment](https://github.com/turbonomic/kubeturbo/pull/655#discussion_r847272590)

**Comment by irfanurrehman** [comment]
> @chlam4 @enlinxu Addressed the review comments. PTAL!
> 
> Additional to the comments posted here, I did address 2 more comments from our online review:
> 
> 1. Introduced a `GitopManager` interface and `GithubManager` implements the same
> 2. Grouped the gitops related files in a gitops specific sub folder. (cc @ading1977 )

@chlam4 @enlinxu Should I merge this?

**Comment by irfanurrehman** [comment]
Rebased on latest!

**Comment by irfanurrehman** [comment]
Have added a feature flag named `GitopsApps` as an alpha feature disabled by default.
It disables only the argocd application discovery and gitops pipeline action execution. The k8s applications will still be discovered as before.

To enable the same below section will need to be added in the kubeturbo config.

```
      "featureGates": {
        "GitopsApps": true
     }
```

@enlinxu @pankajbatra1 This update changes only the last commit. Code prior to that has been reviewed and updated. Please check the last commit and if you are satisfied with the code, please feel free to merge.
cc @chlam4

**Review by pankajbatra1 — APPROVED** [review]

**Pull request merged**